aries_vcx_core/src/anoncreds/credx_anoncreds.rs

```diff
@@ -731,27 +731,59 @@ impl BaseAnonCreds for IndyCredxAnonCreds {
                     format!("Unknown referent: {}", reft),
                 ))?;
 
-            let _attr_name = requested_val.try_get("name")?;
-            let _attr_name = _attr_name.try_as_str()?;
-            let attr_name = _normalize_attr_name(_attr_name);
+            if let Some(_attr_names) = requested_val.get("names").and_then(|v| v.as_array()) {
+                let _attr_names = requested_val.try_get("names")?.as_array().ok_or_else(|| {
+                    AriesVcxCoreError::from_msg(
+                        AriesVcxCoreErrorKind::InvalidAttributesStructure,
+                        "Invalid Json Parsing of Requested Attributes Retrieved From Libindy",
+                    )
+                })?;
 
-            let non_revoked = requested_val.get("non_revoked"); // note that aca-py askar fetches from proof_req json
-            let restrictions = requested_val.get("restrictions");
+                let mut credentials_json = vec![];
 
-            let credx_creds = self
-                ._get_credentials_for_proof_req_for_attr_name(restrictions, &attr_name)
-                .await?;
+                for _attr_name in _attr_names {
+                    let _attr_name = _attr_name.try_as_str()?;
+                    let attr_name = _normalize_attr_name(_attr_name);
 
-            let mut credentials_json = vec![];
+                    let non_revoked = requested_val.get("non_revoked"); // note that aca-py askar fetches from proof_req json
+                    let restrictions = requested_val.get("restrictions");
 
-            for (cred_id, credx_cred) in credx_creds {
-                credentials_json.push(json!({
-                    "cred_info": _make_cred_info(&cred_id, &credx_cred)?,
-                    "interval": non_revoked
-                }))
-            }
+                    let credx_creds = self
+                        ._get_credentials_for_proof_req_for_attr_name(restrictions, &attr_name)
+                        .await?;
 
-            cred_by_attr[ATTRS][reft] = Value::Array(credentials_json);
+                    for (cred_id, credx_cred) in credx_creds {
+                        credentials_json.push(json!({
+                            "cred_info": _make_cred_info(&cred_id, &credx_cred)?,
+                            "interval": non_revoked
+                        }))
+                    }
+                }
```

Comment:
Hmm, I haven't checked how this is done in other implementations, but it seems like something might be missing. If verifier specified `names: ["name", "surname"]` it means that fields from which these fields are selected from are originating from a single credential https://hyperledger.github.io/anoncreds-spec/#presentation-request-example
eg. not mixing "name" from one credential and "surname" from another.

While this code could coincidentally (and in practice currently even likely) work if there's not more alternative credentials in the wallet, it seems like prover is not trying to assure that property.

We don't necessarily have to address this right now, but lets at least create issue to track this.

Reply:
ye IMO this is the whole point of `names`; requiring that multiple attributes are presented from the same credential. `credentials_json` should only be populated with credentials which have ALL the attributes (names) present

Reply:
Hmm no that doesn't fix it. Consider a request for `names: ["a", "b", "c"]`, and a wallet with `cred1 { a, d }`, `cred2 { b, e }`, and `cred3 { c, f }`.

The tweaked implementation with `all_found` would result in `cred1`, `cred2` and `cred3` being added to `credentials_json` and included. Where really, none of the credentials satisfy the requirement of having all 3 attribute names (a, b, and c), so the credentials for that referent should be empty.

A credential should only be added to the list if it has ALL the `name`/`names` as attributes on the credential, AND it satisfies any restrictions applied to it. 

The ideal way to do that IMO is to rewrite `_get_credentials_for_proof_req_for_attr_name` a bit, such that it only needs to be called once per referent, and it takes in an array of `attr_names`, and the WQL query will search for credentials which have ALL those attributes (currently it just checks for 1 attr_name passed in).

Reply:
Anyway, will leave it up to you guys whether that is scope for this PR or for sometime else; as you've stated, the impl is far from perfect and a rewrite is probs best... the original impl was meant to act as a 'base' for the sake of getting it in there, so it's great that you guys are readdress the impl and improving

Reply:
Okay I had another attempt at fixing this. Your insights were very valuable, thank you! I hope the query composition is fine, but I assume we'll see in the CI.